### DISCUSSION TITLE
Slimmer `package.json` for smaller production install

### DISCUSSION THREAD

**Discussion by robak86** [withastro/roadmap#655] [Proposal]
### What version of `astro` are you using?

2.9.3

### Are you using an SSR adapter? If so, which one?

Node

### What package manager are you using?

npm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

After installing only production packages (`npm i --omit dev`) for the project, I noticed that possibly many installed packages are not required for the runtime. They probably should be declared as `devDependencies`. e.g. `typescript`, `@babel/*`, `esbuild`, `vite`. Having them declared as `dependencies` greatly increases the size of `node_modules`, making, e.g. building a small docker image of the app very difficult. The `node_modules` for a "hello world" app has about 160 MB. 

### What's the expected result?

All dependencies that are not required for the runtime should be declared as `devDependencies`.

### Link to Minimal Reproducible Example

https://github.com/withastro/astro/blob/main/packages/astro/package.json

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by Princesseuh** [comment]
Those dependencies are all required by `astro` at runtime. `typescript` is used for `astro check`, for instance, so they can't be marked as `devDeps` or you won't get them when installing `astro`

What you probably mean is that those dependencies are not required by apps built with `astro`, which is a different problem that's not really related to our `package.json` per se

**Comment by robak86** [comment]
Thanks for the clarification. I understand that dependencies like `typescript` are required by the `Astro` command line app. But you don't use `astro check` in a project that will be deployed after running `astro build`. So having, e.g. `typescript` in `devDependencies` would still allow users to run `Astro` CLI in development but wouldn't add additional size to `node_modules` for the production. The dockerization workflow that I'm thinking about is:

1. Builder stage
- copy the project to the Docker context (except for the `node_modules`)
- install all packages (including `devDependencies`)
- run `astro build` - CLI works just fine having `typescript` in `devDependencies.` 

2. Release stage
- copy the `dist` directory from the builder stage
- install **only** production `dependencies` omitting `devDependencies`. No need for installing e.g. `typescript` as we will run only `node dist/server/entry.mjs`

Currently, the last step will also install many dependencies that are not required for `node dist/server/entry.mjs` to run.
However, I'm not sure if this is also the case for other adapters.

**Comment by Princesseuh** [comment]
> So having, e.g. typescript in devDependencies would still allow users to run Astro CLI in development but wouldn't add additional size to node_modules for the production.

No, if `typescript` was a dev dep of `astro`, when users install `astro` in their project they wouldn't get the dev deps

**Reply by rburgst** [reply]
as 99% of people are using the code generator anyway, you could just solve this in the code generator. People with existing repos will need to add typescript as a dev dependency

**Reply by Princesseuh** [reply]
That's how it works nowadays! Astro doesn't have a dependency on TypeScript anymore.

**Comment by robak86** [comment]
Yes, you are right. I forgot that `devDependencies` of 3rd party libraries are not installed. 
Solving this issue would probably require more significant changes - like extracting CLI into a separate package that the user can add as `devDependency`. 
Sorry for the trouble.

**Comment by Princesseuh** [comment]
Oh, no need to close the issue! This is definitely something we'd want to fix eventually. 

I think the best way would probably to actually extrude the runtime part into a separate package and when generating the SSR build, have a package.json with only that package?

**Comment by natemoo-re** [comment]
I'm transferring this to the roadmap because we'll have to talk about it! I think we'd like to see this happen but exactly how we tackle it remains to be seen.

**Comment by natemoo-re** [comment]
Remix uses an interesting pattern here where `@remix-run/dev` is in `devDependencies` but `@remix-run/server` is a production dependency. Pretty nice, tbh.

**Reply by Princesseuh** [reply]
How hard would it be for us to split all our dev code into a separate package? cc @ematipico who's working on refactors recently

**Reply by ematipico** [reply]
I think it would require some work to achieve that, but it's achievable! 

Theoretically, all our runtime code lives inside `src/runtime/`, so it shouldn't be too much work to move more runtime code there (one example that comes to mind is `astro/middleware`). The rest of the code should be considered a dev dependency.

**Comment by firxworx** [comment]
Yeah I am noticing Astro SSR into a docker image is like 700mb-1gb for a "hello world" app with a few things like tailwind + react.

I didn't expect that for all the efficiency and optimization of Astro -- which I love -- when it comes to deployment it insists on the kitchen sink of `node_modules/` or else problems abound.

This makes it a challenge to work with deployments and blows up costs and absolutely impacts performance. Consider for example a serverless container (e.g. AWS Lambda Container Functions) the cold-start time can be blown up by having a needlessly large image.

In a monorepo with common dependency versions (e.g. nx with a common "base" package.json) then Astro is _really_ a challenge to deploy.

If the tooling could produce the dependency graph required for deployment or generate a package.json that would be amazing.
(and in the most ideal but perhaps further-off case: lock files for the leading package managers).

Any ideas on how Astro projects (especially SSR) don't have to be the size of a node_modules black hole on deploy?

**Comment by linguofeng** [comment]
👀

**Comment by MoltenCoffee** [comment]
I manage to reduce my docker builds greatly by moving a lot of stuff into `devDependencies`.
`astro` can safely be put in there, as well as other packages that pull in `typescript`, `prettier`, you name it. If needed, these can then be added in `astro.config.mjs` like so, so they are tree-shaken and bundled:

```js
import { defineConfig} from "astro/config";

export default defineConfig({
  vite: {
    ssr: {
      noExternal: ["react-email"],
    },
  },
});
```

Dockerfile:

```
FROM node:24.21.0-alpine AS base
WORKDIR /app
RUN corepack enable

ENV ASTRO_TELEMETRY_DISABLED=1

COPY package.json pnpm-lock.yaml pnpm-workspace.yaml ./

FROM base AS prod-deps
RUN pnpm install --prod --frozen-lockfile

FROM base AS build-deps
RUN pnpm install --frozen-lockfile

FROM build-deps AS build
COPY . .
RUN npm run build

FROM base AS runtime
ENV NODE_ENV=production
ENV HOST=0.0.0.0
ENV PORT=4321

COPY --from=prod-deps /app/node_modules ./node_modules
COPY --from=build /app/dist ./dist

EXPOSE 4321
CMD ["node", "./dist/server/entry.mjs"]
```
